# Kenya Agriculture Data Analytics

## Data Loading

This notebook demonstrates how to load the Kenya Agriculture dataset into Apache Spark for distributed data processing.

**Objectives**
- Create a Spark session
- Load the CSV dataset
- Inspect the data structure
- Verify data quality before transformation

In [16]:
import os
import sys

# Get the project root (parent of notebooks/)
project_root = os.path.abspath("..")

# Add it to Python's search path
if project_root not in sys.path:
    sys.path.insert(0, project_root)

print(project_root)

c:\Users\ADMIN\OneDrive\Desktop\OUK Msc\1.2\Data Mining and Big Data\Big-Data-Analytics-with-Hadoop-and-Spark


In [17]:
import os

print(os.path.exists("../scripts"))
print(os.listdir("../scripts"))

True
['analysis.py', 'load_data.py', 'main.py', 'transformations.py', '__pycache__']


In [19]:
from scripts.load_data import create_spark_session, load_dataset

spark = create_spark_session()

df = load_dataset(spark, "../data/kenya_agriculture_raw.csv")

df.show(5)

+-----------+--------------------+---------------+-----+------------+----------+---------------+--------------------+---------+----+------+-----+----+----------------+
|Domain Code|              Domain|Area Code (M49)| Area|Element Code|   Element|Item Code (CPC)|                Item|Year Code|Year|  Unit|Value|Flag|Flag Description|
+-----------+--------------------+---------------+-----+------------+----------+---------------+--------------------+---------+----+------+-----+----+----------------+
|        QCL|Crops and livesto...|            404|Kenya|        5510|Production|        1929.07|Abaca, manila hem...|     1976|1976|tonnes| 10.0|   E| Estimated value|
|        QCL|Crops and livesto...|            404|Kenya|        5510|Production|        1929.07|Abaca, manila hem...|     1977|1977|tonnes| 10.0|   E| Estimated value|
|        QCL|Crops and livesto...|            404|Kenya|        5510|Production|        1929.07|Abaca, manila hem...|     1978|1978|tonnes| 10.0|   E| Estimated

In [20]:
df.printSchema()

root
 |-- Domain Code: string (nullable = true)
 |-- Domain: string (nullable = true)
 |-- Area Code (M49): integer (nullable = true)
 |-- Area: string (nullable = true)
 |-- Element Code: integer (nullable = true)
 |-- Element: string (nullable = true)
 |-- Item Code (CPC): double (nullable = true)
 |-- Item: string (nullable = true)
 |-- Year Code: integer (nullable = true)
 |-- Year: integer (nullable = true)
 |-- Unit: string (nullable = true)
 |-- Value: double (nullable = true)
 |-- Flag: string (nullable = true)
 |-- Flag Description: string (nullable = true)



In [21]:
print(df.count())

18182


In [22]:
print(df.columns)

['Domain Code', 'Domain', 'Area Code (M49)', 'Area', 'Element Code', 'Element', 'Item Code (CPC)', 'Item', 'Year Code', 'Year', 'Unit', 'Value', 'Flag', 'Flag Description']


In [23]:
df.describe().show()

+-------+-----------+--------------------+---------------+-----+-----------------+--------------------+------------------+--------------------+------------------+------------------+-------+------------------+-----+-----------------+
|summary|Domain Code|              Domain|Area Code (M49)| Area|     Element Code|             Element|   Item Code (CPC)|                Item|         Year Code|              Year|   Unit|             Value| Flag| Flag Description|
+-------+-----------+--------------------+---------------+-----+-----------------+--------------------+------------------+--------------------+------------------+------------------+-------+------------------+-----+-----------------+
|  count|      18182|               18182|          18182|18182|            18182|               18182|             18182|               18182|             18182|             18182|  18182|             18182|18182|            18182|
|   mean|       NULL|                NULL|          404.0| NULL|5413

In [24]:
from pyspark.sql.functions import col, sum

df.select([
    sum(col(c).isNull().cast("int")).alias(c)
    for c in df.columns
]).show()

+-----------+------+---------------+----+------------+-------+---------------+----+---------+----+----+-----+----+----------------+
|Domain Code|Domain|Area Code (M49)|Area|Element Code|Element|Item Code (CPC)|Item|Year Code|Year|Unit|Value|Flag|Flag Description|
+-----------+------+---------------+----+------------+-------+---------------+----+---------+----+----+-----+----+----------------+
|          0|     0|              0|   0|           0|      0|              0|   0|        0|   0|   0|    0|   0|               0|
+-----------+------+---------------+----+------------+-------+---------------+----+---------+----+----+-----+----+----------------+



## Data Loading Summary

The dataset was successfully loaded into PySpark and inspected to understand its structure and quality before any transformations.

### Dataset Structure

The schema shows that the dataset contains **14 columns** describing agricultural production in Kenya. These include identifiers (Domain Code, Item Code, Element Code), descriptive fields (Domain, Area, Item, Element), temporal information (Year, Year Code), measurement units, production values, and data quality flags.

The inferred data types indicate that:
- Categorical variables are stored as strings.
- Numeric identifiers are stored as integers or doubles.
- The **Value** column is numeric and will be used for analysis.

### Dataset Size

The dataset contains **18,182 records**, providing sufficient historical agricultural data for analysis.

### Column Names

The dataset consists of the following columns:

- Domain Code
- Domain
- Area Code (M49)
- Area
- Element Code
- Element
- Item Code (CPC)
- Item
- Year Code
- Year
- Unit
- Value
- Flag
- Flag Description

These columns describe agricultural commodities, production metrics, geographical information, time periods, measurement units, and metadata about data quality.

### Descriptive Statistics

The summary statistics indicate that:

- All **18,182 records** contain valid values for every column.
- The data spans from **1961 to 2021**, covering approximately 60 years of agricultural production.
- The **Value** column ranges from **0** to approximately **36 million**, indicating substantial variation between agricultural products and production levels.
- Kenya is the only geographical area represented in the dataset (Area Code 404).
- The dataset contains multiple agricultural commodities measured in different units.

### Missing Value Assessment

A missing value check was performed across all columns.

The results show that **no missing values** were found in any column. This indicates that the dataset is complete and does not require missing-value imputation before the data transformation stage.

### Conclusion

The data loading phase confirms that the dataset was imported successfully, the schema was inferred correctly, all expected columns are present, and the dataset contains complete records without missing values. The dataset is therefore ready for the next stage of the pipeline, which involves data cleaning and transformation.